ex7/t_hayakawa/main.py

```diff
@@ -0,0 +1,329 @@
+        pi, mu, sig, lh_list, iter = gmm.iteration(data, I=100, e=0.01)
+
+        calc_time = time.time() - start
+        print(f"calc_time({method}): {calc_time:.5f}")
+
+        xx = np.linspace(np.min(data[:, 0]), np.max(data[:, 0]), n)
+        pdfs = np.zeros((n, K))
+        for k in range(K):
+            pdfs[:, k] = pi[k] * norm.pdf(xx, loc=mu[k], scale=sig[k])
+
+        fig = plt.figure(figsize=(8, 6))
+        plt.subplots_adjust(hspace=0.6)
+        fig.add_subplot(
+            211,
+            title=f"{args.filename} GMM ({iter} iter)",
+            xlabel="x",
+            ylabel="probability",
+        )
+        plt.scatter(
+            data,
+            np.zeros(data.shape[0]),
+            facecolor="None",
+            edgecolors="darkblue",
+            label="data",
+        )
+        plt.scatter(
+            mu, np.zeros(mu.shape[0]), s=100, marker="x", c="r", label="centroid",
+        )
+        plt.plot(xx, np.sum(pdfs, axis=1), label="GMM")
+        plt.legend()
+
+        fig.add_subplot(
+            212,
+            title=f"{args.filename} GMM stack ({iter} iter)",
+            xlabel="x",
+            ylabel="probability",
+        )
+        labels = []
+        stacks = np.zeros((0, pdfs.shape[0]))
+        for k in range(K):
+            labels.append(f"$\pi{k}$ : {pi[k]:.3f}")
+            stacks = np.vstack((stacks, pdfs[:, k]))
+        plt.stackplot(xx, stacks, labels=labels)
+        plt.legend(bbox_to_anchor=(0.25, 1))
+        plt.savefig(f"{args.filename}_gmm.png")
+        plt.show()
+
+        # 対数尤度関数
+        fig = plt.figure()
+        fig.add_subplot(
+            111,
+            title=f"{args.filename} Likelihood",
+            xlabel="Iteration",
+            ylabel="Log Likelihood",
+        )
+        plt.plot(lh_list[:, 0], lh_list[:, 1])
+        plt.savefig(f"{args.filename}_lh.png")
+        plt.show()
+
+    elif data.shape[1] == 2:
+        start = time.time()
+
+        if method == "r":
+            mu_x = (np.max(data[:, 0]) - np.min(data[:, 0])) * np.random.rand(
+                K, 1
+            ) + np.min(data[:, 0])
+            mu_y = (np.max(data[:, 1]) - np.min(data[:, 1])) * np.random.rand(
+                K, 1
+            ) + np.min(data[:, 1])
+            mu = np.hstack((mu_x, mu_y))
+
+            # 単位行列 * 正の乱数
+            sig = np.random.rand(K, 2, 2) * np.identity(2)
+
+            pi = np.zeros(K)
+            pi += 1 / K
+        elif method == "km":
+            model = KMeans(n_clusters=K, init="random")
+            fit_data = model.fit_predict(data)
+            mu = model.cluster_centers_
+
+            # 共分散
+            index_data = [np.where(fit_data == i) for i in range(K)]
+            sig = [[np.cov(data[index_data[i]].T)] for i in range(K)]
+            sig = np.squeeze(sig)
+
+            # 混合係数
+            pi = np.zeros(K)
+            for k in range(K):
+                pi[k] = index_data[k][0].shape[0] / data.shape[0]
```

Comment:
ちなみに for なしで頑張って書くとこうなります。muの分も一応書いてます。正直あってるか少し不安です。修正は不要です
参考：[numpy.add.at](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.at.html)
```suggestion
            n = np.zeros(K)
            # n(K,) -> n[fit_data](N,)
            # n[fit_data](N,) += 1
            np.add.at(n, fit_data, 1)
            pi = n / data.shape[0]
            # mu(K,D) -> mu[fit_data](N,D)
            # mu[fit_data](N,D) += data(N,D)
            np.add.at(mu, fit_data, data)
            mu /= n[:,np.newaxis]
            # mu(K,D) -> mu[fit_data](N,D)
            # tmp(N,D) = data(N,D) - mu[fit_data](N,D)
            tmp = data - mu[fit_data]
            # tmp(N,D,D) = tmp[:,:,np.newaxis](N,D,1) * tmp[:,np.newaxis,:](N,1,D)
            tmp = tmp[:,:,np.newaxis] * tmp[:,np.newaxis,:]
            # sig(K,D,D) -> sig[fit_data](N,D,D)
            # sig[fit_data](N,D,D) += tmp(N,D,D)
            np.add.at(sig, fit_data, tmp)
            sig /= n[:,np.newaxis,np.newaxis]
```